relay-server/src/actors/project_upstream.rs

```diff
@@ -130,36 +188,34 @@ impl UpstreamProjectSource {
         self.config.query_batch_interval() + self.backoff.next_backoff()
     }
 
-    /// Schedules a batched upstream query with exponential backoff.
-    fn schedule_fetch(&mut self, context: &mut Context<Self>) {
-        utils::run_later(self.next_backoff(), Self::fetch_states).spawn(context)
-    }
-
     /// Executes an upstream request to fetch project configs.
     ///
     /// This assumes that currently no request is running. If the upstream request fails or new
     /// channels are pushed in the meanwhile, this will reschedule automatically.
-    fn fetch_states(&mut self, context: &mut Context<Self>) {
-        if self.state_channels.is_empty() {
-            relay_log::error!("project state update scheduled without projects");
-            return;
-        }
-
-        let batch_size = self.config.query_batch_size();
-        let num_batches = self.config.max_concurrent_queries();
+    async fn fetch_states(
+        config: Arc<Config>,
+        mut state_channels: HashMap<ProjectKey, ProjectStateChannel>,
+        inner_tx: mpsc::UnboundedSender<(
+            HashMap<ProjectKey, ProjectStateChannel>,
+            ProjectStatesResponses,
+        )>,
+        attempt: usize,
+    ) {
+        let batch_size = config.query_batch_size();
+        let num_batches = config.max_concurrent_queries();
 
         // Pop N items from state_channels. Intuitively, we would use
-        // `self.state_channels.drain().take(n)`, but that clears the entire hashmap regardless how
+        // `state_channels.drain().take(n)`, but that clears the entire hashmap regardless how
         // much of the iterator is consumed.
         //
         // Instead, we have to collect the keys we want into a separate vector and pop them
         // one-by-one.
-        let projects: Vec<_> = (self.state_channels.keys().copied())
+        let projects: Vec<_> = (state_channels.keys().copied())
             .take(batch_size * num_batches)
```

Comment:
If you move this `take` to `schedule_fetch`, you ensure that all channels that won't be fetched remain in the map and get appended to.

I would even go further and also move:
 - The `filter_map` + `filter` below, to further reduce allocations of the map that's being moved into the background job
 - The metric and `debug!` log, since those need the full size of the map. 

In a way, that splits the old `fetch_states` more into two functions, but the semantics of these two blocks match:
- `schedule_fetch` selects which states to fetch
- `fetch_states` only runs the fetch

Reply:
move this part into the separate function [462a093](https://github.com/getsentry/relay/pull/1727/commits/462a09371fb1cfb026641ae515672790b4b1c593)